accepted/future-releases/static-extension-methods/feature-specification.md

```diff
@@ -46,214 +46,205 @@ extension MyFancyList<T> on List<T> {
 
-```dart
-extension<T> on List<T> {
-  void quadruple() { ... }
-}
-```
+An extension declaration with a non-private name is included in the library's export scope, and a privately named extension is not. It is a compile-time error to export two declarations, including extensions, with the same name, whether they come from declarations in the library itself or from export declarations (with the usual exception when all but one declaration come from platform libraries). Extension *members* with private names are simply inaccessible in other libraries.
 
-This is equivalent to giving the extension a fresh private name.
+We make `extension` a built-in identifier. Is not necessary for disambiguation, but it makes error-recovery in parsers much easier.
 
 If we make `on` a built-in identifier, then there should not be any parsing issue. Even without that, the grammar should be unambiguous because `extension on on on { … }` and `extension on on { … }` are distinguishable, and the final type cannot be empty. It may be *harder* to parse.
 
-This is a simple feature, but with very low impact. It only allows you to omit a single private name for an extension that is only used in a single library.
+The ability to implicitly give an extension a private name is a simple feature, but with very low impact. It only allows you to omit a single private name for an extension that is only used in a single library.
 
-### Scope
+### Explicit Extension Member Invocation
 
-Dart's static extension methods are *scoped*. They only apply to code where the extension itself is *in scope*. Being in scope means that the extension is declared or imported into a scope which is a parent scope of the current lexical scope.
+You can explicitly invoke an extension member on a particular object by performing a *member invocation* on an *extension application*.
 
-You can *avoid* making the extension in-scope for a library by either not importing any library exporting the extension, importing such a library and hiding the extension using `hide` or `show`, or importing such a library only with a prefix.
+An *extension application* is an expression of the form `E(expr)` or `E<typeArgs>(expr)` where `E` denotes an extension declaration (that is, `E` a simple or qualified identifier which refers to the extension declaration).
 
-An extension *is* in scope if the name is *shadowed* by another declaration (a class or local variable with the same name shadowing a top-level or imported declaration, a top-level declaration shadowing an imported extension, or a non-platform import shadowing a platform import).
+An extension application is subject to static type inference.  If `E` is an extension declared as `extension E<X...> on T {...}`, then the type inference for an extension application is exactly the same as that the type inference performed for the same syntax considered as a constructor invocation on a class declared as:
```

Comment:
Does the context type ever affect the receiver expression of `E(o).m`? I was assuming that there is no context type for `E(o)` in that expression because there is no way to pass a context type through an application of something that we don't know what is (yet). Then, after `E(o)` was inferred with no context type, we use the context type and type of `E(o)` to do inference on the `.m` invocation.

Reply:
For direct property accesses, there is never a context type.  For cascades, there may be.  So if this doesn't apply to cascades, then this is implicit, but I think it does?